Always Encrypted

fredwes opened the pull request:
Add Always Encrypted support to tedious.js.

Done PR submission steps 1-6

fredwes committed:
feat: always encrypted

#447

akshayganeshen committed:
test: always encrypted

#447

akshayganeshen committed:
chore: fix lint issues

arthurschreiber commented:
Oh, wow. This is quite the Christmas present! 🎄 🎁

arthurschreiber commented:
So I took a quick look over the changes - there is quite a lot of changes here! How long have you been working on this? 😲 

Anyway, it'll take me a bit of time to have this reviewed and give feedback for this, but the changes are looking pretty solid from what I've seen so far. 👍

fredwes commented:
We've worked on it for a few months but only recently got approval to submit a PR. Would love to get your feedback! 😄

IanChokS commented:
Hi @fredwes,

Thanks for this PR! Just wondering if this PR also allows for keys to be stored in Windows Certificate Store?

fredwes commented:
@IanChokS The only key store supported by this PR is Azure Key Vault. Additional key stores can be added by implementing a `decryptColumnEncryptionKey` method

fredwes committed:
refactor: add tests and clean up code

fredwes committed:
fix: var char test

fredwes committed:
chore: fix lint issues

fredwes committed:
test: fix always-enrypted integration test

fredwes committed:
Merge remote-tracking branch 'upstream/master'

arthurschreiber commented on src/always-encrypted/cek-table.ts:
```diff
@@ -0,0 +1,62 @@
+import { EncryptionKeyInfo } from './types';
+
+export class CEKTable {
```
Do we need this class, or could we just use an array instead to hold CEK values?

fredwes replied:
It's not needed. I've removed it and replaced it with an array as you suggested

arthurschreiber commented:
@fredwes Thanks for all the work on this so far! I added some comments around some of the code that I reviewed.

fredwes committed:
fix: max callstack exceeded in stream-parser

fredwes committed:
test: fix always-encrypted test error handling

fredwes committed:
test: skip always-encrypted integ test

Do not run integ test on unsupported TDS versions.

fredwes committed:
refactor: implement code review comments

fredwes committed:
chore: fix lint issues

fredwes committed:
chore: fix azure lib typescript errors

fredwes commented:
Thanks @arthurschreiber! I fixed the issues you raised so far and fixed the rest of the failing tests. I'm going to continue to add additional test cases as you complete your review

fredwes committed:
refactor: address more PR comments

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+
+export const algorithmName = 'AEAD_AES_256_CBC_HMAC_SHA256';
+
+export class AeadAes256CbcHmac256Algorithm implements EncryptionAlgorithm {
```
I updated the relevant files to include the copyright notice.

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+    this.versionSize = Buffer.from([1]);
+    this.minimumCipherTextLengthInBytesNoAuthenticationTag = 1 + this.blockSizeInBytes + this.blockSizeInBytes;
+    this.minimumCipherTextLengthInBytesWithAuthenticationTag = this.minimumCipherTextLengthInBytesNoAuthenticationTag + this.keySizeInBytes;
+    this.algorithmVersion = Buffer.from([0x1]);
```
Good idea, I've moved it out as a constant

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+    if (encryptionType === SQLServerEncryptionType.Deterministic) {
+      this.isDeterministic = true;
+    }
+    this.version = Buffer.from(this.algorithmVersion);
```
Fixed

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+    this.algorithmVersion = Buffer.from([0x1]);
+    this.columnEncryptionkey = columnEncryptionKey;
+
+    if (encryptionType === SQLServerEncryptionType.Deterministic) {
```
Done

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+
+  constructor(columnEncryptionKey: AeadAes256CbcHmac256EncryptionKey, encryptionType: SQLServerEncryptionType) {
+    this.isDeterministic = false;
+    this.blockSizeInBytes = 16;
```
Good idea, I've moved it out as a constant

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+    let iv: Buffer;
+
+    if (this.isDeterministic === true) {
+      try {
```
You're right, I don't think it is necessary

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+      if (numBlocks > 1) {
+        count = (numBlocks - 1) * this.blockSizeInBytes;
+
+        encryptCipher.update(plaintText.slice(0, count)).copy(outBuffer, cipherIndex, 0, count);
```
This was ported over from `mssql-jdbc`, but doesn't seem to be required. I've removed this logic

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+        const hmac = createHmac('sha256', this.columnEncryptionkey.getMacKey());
+        hmac.update(this.version);
+        hmac.update(iv);
+        hmac.update(outBuffer.slice(cipherStartIndex, (numBlocks * this.blockSizeInBytes) + cipherStartIndex));
```
Good idea, I've refactor it now

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+  }
+
+  decryptData(cipherText: Buffer): Buffer {
+    const hasAuthenticationTag = true;
```
This was ported over from mssql-jdbc, but doesn't seem to be required. I've removed this logic

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+    return plainText;
+  }
+
+  _prepareAuthenticationTag(iv: Buffer, cipherText: Buffer, offset: number, length: number): Buffer {
```
Done

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-algorithm.ts:
```diff
@@ -0,0 +1,170 @@
+    hmac.update(this.versionSize);
+    const computedHash = hmac.digest();
+
+    computedHash.copy(authenticationTag, 0, 0, authenticationTag.length);
```
You are correct. Fixed.

fredwes commented on src/always-encrypted/aead-aes-256-cbc-hmac-encryption-key.ts:
```diff
@@ -0,0 +1,70 @@
+export const deriveKey = (rootKey: Buffer, salt: string): Buffer => {
+  const hmac = createHmac('sha256', rootKey);
+  hmac.update(Buffer.from(salt, 'utf16le'));
+  const derivedKeyBuff = hmac.digest().slice(0, keySizeInBytes);
```
Fixed

fredwes commented on src/always-encrypted/key-crypto.ts:
```diff
@@ -0,0 +1,107 @@
+  const algorithmName = validateAndGetEncryptionAlgorithmName(md.cipherAlgorithmId, md.cipherAlgorithmName);
+  const cipherAlgorithm = new AeadAes256CbcHmac256Algorithm(new AeadAes256CbcHmac256EncryptionKey(symKey.rootKey, algorithmName), md.encryptionType);
+
+  if (!cipherAlgorithm) {
```
Removed

fredwes committed:
test: fix varcharMaxKnownLengthWrong test

fredwes committed:
Merge remote-tracking branch 'upstream/master'

fredwes committed:
chore: fix lint and merge issues

arthurschreiber commented on tsconfig.json:
```diff
@@ -8,7 +8,7 @@
     "esModuleInterop": true,
     "allowSyntheticDefaultImports": true,
-    "lib": [ "esnext" ]
+    "lib": [ "esnext", "DOM" ]
```
@fredwes Where does this change come from? Node.js does not really expose the DOM libraries, so I'm wondering why this is required. 🤔

fredwes replied:
@arthurschreiber this is because the Azure libs for the Azure key vault reference some types from DOM, TS will complain that it can't find those standard types without it

e.g.: the "Headers" type here: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-http/lib/fetchHttpClient.ts#L198

arthurschreiber commented:
@fredwes We're currently in the process of changing how requests and their payloads are generated - see https://github.com/tediousjs/tedious/issues/1038.

I believe those changes will nicely play into some of the changes you've done for the encryption support (e.g. adding `toBuffer`). That will require some updates to the changes you've done here so far, but also should reduce a whole bunch of duplication.

What would be absolutely awesome is if we could break out some of the changes made in here in support for Always Encrypted but which are not really related to encryption support. What I'm mainly thinking about is the `collation` support for parameters. Extracting that into a separate PR would allow me to review and merge this separately from the larger PR here, while also making both PRs easier to review in the end. If you don't have time for doing this yourself, would you be ok if I do the extraction?

What do you think? 🙇

fredwes commented:
@arthurschreiber That's a good idea, I think there are some other things that can be split out into separate PRs as well (and cleaned up). I can do it or if you prefer to do it, that's fine too. I can get started in the next couple days.

arthurschreiber commented:
@fredwes If I understand correctly, Always encryoption does not support PLP streamed data (e.g. for `nvarchar(max)` or `varbinary(max)`).

Does this mean encrypted data is limited to `4000` characters / `8000` bytes? 🤔

fredwes commented:
@arthurschreiber Always Encrypted does support PLP, but for the purpose of `toBuffer` we don't use PLP terminators, the _entire_ parameter value is used during encryption/decryption

arthurschreiber commented:
@fredwes how would a PLP implementation look like? I’m wondering because the authentication tag seems to be stored at the front of the encrypted data, but can only be generated after reading all data, making PLP streaming pretty useless, I’d think. 🤔

fredwes commented:
@arthurschreiber There is no implementation necessary from an Always Encrypted perspective. It should be supported implicitly using existing parameter writing and reading logic. You are right, PLP streaming for Always Encrypted is essentially just a compatibility thing, since you need the entire value to do encrypt/decrypt.

Sathishabm210 commented:
@fredwes  Here is my code to retrieve the always encrypted column data using tedious-master, still I am getting encrypted data only. Will this support to get decrypted data from the always encrypted column table?
Below is my code:

```
var Connection = require('tedious-master').Connection;
var ColumnEncryptionAzureKeyVaultProvider = require('tediousmaster').ColumnEncryptionAzureKeyVaultProvider;
var Request = require('tedious-master').Request;

```
```
var config = {  
    server: '',  //update me
    authentication: {
        type: 'default',
        options: {
            userName: '', //update me
            password: ''  //update me
        }
    },
    options: {
        // If you are on Microsoft Azure, you need encryption:
        encrypt: true,
        database: ''  //update me
    }
};
```

```
var akvProvider = new ColumnEncryptionAzureKeyVaultProvider("", "", "");
config.options.encryptionKeyStoreProviders = [{
  key: akvProvider.name,
  value: akvProvider
}]
```
```
var connection = new Connection(config);
connection.on('connect', function(err) {  
    // If no error, then good to proceed.  
    console.log("Connected");  
    executeStatement();  
});  

```
```
function executeStatement() {  
    request = new Request("select * from alwaysEncryptColumnTable ", function(err) {  
    if (err) {  
        console.log(err);}  
    });  
    var result = "";  
    request.on('row', function(columns) {  
        columns.forEach(function(column) {  
          if (column.value === null) {  
            console.log('NULL');  
          } else {  
            result+= column.value + " ";  
          }  
        });  
        console.log(result);  
        result ="";  
    });  
    request.on('done', function(rowCount, more) {  
    console.log(rowCount + ' rows returned');  
    });  
    connection.execSql(request);  
}  
```

fredwes commented:
@Sathishabm210 are you using tedious master branch or this PR branch? Always Encrypted feature isn’t merged yet to master. 

If you are using this PR branch then I see that you are missing the Always Encrypted connection option. Add this property to your connection options: `alwaysEncryptedSetting: true`

jocund commented:
@fredwes Just tried always encrypted based connection using AKV referencing the example you provided. Was able to push/pull non-encrypted columns, but for encrypted one it says the following for me
**(node:16044) UnhandledPromiseRejectionWarning: Error: Cannot use a non-RSA key: RSA-HSM.**

Does this imply that we cannot set a key vault with RSA-HSM if we need to use your existing code? Any thing we can do? Please let me know

Sathishabm210 commented:
> @Sathishabm210 are you using tedious master branch or this PR branch? Always Encrypted feature isn’t merged yet to master.
> 
> If you are using this PR branch then I see that you are missing the Always Encrypted connection option. Add this property to your connection options: `columnEncryptionSetting: true`

@fredwes I am using this PR branch only. After setting this option its throwing the below error
```
Error: Cannot use a non-RSA key: RSA-HSM.
    at ColumnEncryptionAzureKeyVaultProvider.getAKVKeySize (C:\projects\azureKeyvault.js\node_modules\tedious-master\lib\always-encrypted\keystore-provider-azure-key-vault.js:296:13)
    at C:\projects\azureKeyvault.js\node_modules\tedious-master\lib\always-encrypted\keystore-provider-azure-key-vault.js:51:36
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (C:\projects\azureKeyvault.js\node_modules\tedious-master\lib\always-encrypted\keystore-provider-azure-key-vault.js:16:103)
    at _next (C:\projects\azureKeyvault.js\node_modules\tedious-master\lib\always-encrypted\keystore-provider-azure-key-vault.js:18:194)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
(node:5300) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch
block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:5300) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

fredwes commented:
@Sathishabm210 @jocund Looks like there is a bug with my code. `RSA-HSM` should be allowed.
This PR has gotten out of date with the master branch. I will make sure to apply a fix for this in one of my upcoming PRs.

jocund commented:
@fredwes Thanks for confirming. When can we expect a PR update to fix this? Also, in the mean time can you suggest any work around that we could do on this PR to help move forward?

jocund commented:
@fredwes Commented this line in your lib folder file 'keystore-provider-azure-key-vault' that checks 'RSA' to allow 'RSA-HSM' based keys. It worked to help execute successful pull queries 

// if (!kty || 'RSA'.localeCompare(kty, 'en') !== 0) {
// throw new Error(`Cannot use a non-RSA key: ${kty}.`);

However, when inserting data into encrypted column, I've landed with a 'collation_name' issue as below. Do you know what we can do to resolve it? 

 'Operand type clash: 
 char(6) encrypted with 
 (encryption_type = \'DETERMINISTIC\', 
 encryption_algorithm_name = \'AEAD_AES_256_CBC_HMAC_SHA_256\', 
 column_encryption_key_name = \'Our_key_Name\', 
 column_encryption_key_database_name = \'Our_DB_Name\') 
 collation_name = \'\' 

 is incompatible with char(6) encrypted with (
 encryption_type = \'DETERMINISTIC\', 
 encryption_algorithm_name = \'AEAD_AES_256_CBC_HMAC_SHA_256\', 
 column_encryption_key_name = \'Our_key_Name\', 
 column_encryption_key_database_name = \'Our_DB_Name\') 
 collation_name = \'SQL_Latin1_General_CP1_CI_AS\'',

fredwes commented:
> @fredwes Commented this line in your lib folder file 'keystore-provider-azure-key-vault' that checks 'RSA' to allow 'RSA-HSM' based keys. It worked to help execute successful pull queries
> 
> // if (!kty || 'RSA'.localeCompare(kty, 'en') !== 0) {
> // throw new Error(`Cannot use a non-RSA key: ${kty}.`);
> 
> However, when inserting data into encrypted column, I've landed with a 'collation_name' issue as below. Do you know what we can do to resolve it?
> 
> 'Operand type clash:
> char(6) encrypted with
> (encryption_type = 'DETERMINISTIC',
> encryption_algorithm_name = 'AEAD_AES_256_CBC_HMAC_SHA_256',
> column_encryption_key_name = 'Our_key_Name',
> column_encryption_key_database_name = 'Our_DB_Name')
> collation_name = ''
> 
> is incompatible with char(6) encrypted with (
> encryption_type = 'DETERMINISTIC',
> encryption_algorithm_name = 'AEAD_AES_256_CBC_HMAC_SHA_256',
> column_encryption_key_name = 'Our_key_Name',
> column_encryption_key_database_name = 'Our_DB_Name')
> collation_name = 'SQL_Latin1_General_CP1_CI_AS'',

@jocund That change should fix your issue with RSA-HSM. With regard to the error you are getting, that is an error from SQL server itself. You must use a BIN2 collation with deterministic encryption (see here: https://docs.microsoft.com/en-us/sql/relational-databases/security/encryption/always-encrypted-database-engine?view=sql-server-ver15#selecting--deterministic-or-randomized-encryption).

jocund commented:
@fredwes Thanks. We are using BIN2 with deterministic, but we noticed that the operand type clash error is occurring only for "Char" and "VarChar" data types on SQL. When we change these to "NChar" and "NVarChar" on SQL columns and use those types on our code (using tedious), our insert is successful. Not sure why these specific data types would cause an issue. I did notice that your code has "Collation" logic in Char.js and not in NChar.js - Do you think that's playing a role? 

**char.js** 
 writeTypeInfo: function writeTypeInfo(buffer, parameter) {
    buffer.writeUInt8(this.id);
    buffer.writeUInt16LE(parameter.length);
    const collation = Buffer.alloc(5);

    if (parameter.collation != null) {
      const _parameter$collation = parameter.collation,
            lcid = _parameter$collation.lcid,
            flags = _parameter$collation.flags,
            version = _parameter$collation.version,
            sortId = _parameter$collation.sortId;
      collation.writeUInt8(lcid & 0xFF, 0);
      collation.writeUInt8(lcid >> 8 & 0xFF, 1); // byte index 2 contains data for both lcid and flags

      collation.writeUInt8(lcid >> 16 & 0x0F | (flags & 0x0F) << 4, 2); // byte index 3 contains data for both flags and version

      collation.writeUInt8(flags & 0xF0 | version & 0x0F, 3);
      collation.writeUInt8(sortId & 0xFF, 4);
    }

    buffer.writeBuffer(collation);
  },

**nchar.js** 

  writeTypeInfo: function writeTypeInfo(buffer, parameter) {
    buffer.writeUInt8(this.id);
    buffer.writeUInt16LE(parameter.length * 2);
    buffer.writeBuffer(Buffer.from([0x00, 0x00, 0x00, 0x00, 0x00]));
  },

fredwes commented:
@jocund I think the issue is that you will need to specify the collation in the parameter options.
```javascript
request.addParameter('name', TYPES.Char, 'value', { collation: { sortId: 52, lcid: 0, flags: 0, version: 0 } });
```

IanChokS commented on src/rpcrequest-payload.ts:
```diff
@@ -116,11 +135,81 @@ class RpcRequestPayload {
+          }
+          const plainTextBuffer = type.toBuffer(parameters[i], this.options);
+          if (plainTextBuffer) {
+            promises.push(encryptWithKey(plainTextBuffer, parameters[i].cryptoMetadata as CryptoMetadata, this.options).then((encryptedValue: Buffer) => {
+              parameters[i].encryptedVal = encryptedValue;
+            }));
+          }
+        }
+      }
+
+      Promise.all(promises).then(() => callback(parameters));
```
@fredwes Is there a reason why `encryptWithKey` is wrapped in a promise? If `serverSupportsColumnEncryption = true`, then before the call back is called (on line 189), it would have to wait for all the parameters to be encrypted, and there's nothing else to do in this function in the meantime as it's being encrypted. Wouldn't that just make this process synchronous then?

fredwes replied:
@IanChokS It waits for all parameters to be encrypted before continuing. We saw a significant performance improvement this way. Especially when parameters use different master/column encryption keys (since fetching the keys can happen in parallel this way). It may be worth testing this again with the recent generator refactors though

jocund commented:
@fredwes When we try to pull more than 8-10 Azure Always encrypted records from DB using your code, we're seeing that there is some issue with "unknown type error". See logs below. Would you know how to fix it? Note, when we only pull less that 8 records, the pull is successful 

(node:10452) UnhandledPromiseRejectionWarning: Error: Unknown type: 2
 at C:\Users\Azure_Server\tedious-master\lib\token\stream-parser.js:181:35
 at C:\Users\Azure_Server\tedious-master\lib\tracking-buffer\readable-tracking-buffer.js:67:7
 at ReadableTrackingBuffer.awaitData (C:\Users\Azure_Server\tedious-master\lib\tracking-buffer\readable-tracking-buffer.js:49:7)
 at ReadableTrackingBuffer.readUInt8 (C:\Users\Azure_Server\tedious-master\lib\tracking-buffer\readable-tracking-buffer.js:64:10)
 at C:\Users\Azure_Server\tedious-master\lib\token\stream-parser.js:221:49
 at Parser.awaitData (C:\Users\Azure_Server\tedious-master\lib\token\stream-parser.js:208:7)
 at Parser.readUInt8 (C:\Users\Azure_Server\tedious-master\lib\token\stream-parser.js:221:10)
 at C:\Users\Azure_Server\tedious-master\lib\token\stream-parser.js:174:17
 at new Promise (<anonymous>)
 at C:\Users\Azure_Server\tedious-master\lib\token\stream-parser.js:173:15

(node:10452) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)

jocund commented:
@fredwes Grateful for your urgent assistance for above question. Can you please advice? I'm running the code on Azure app service linux env

akshayganeshen commented:
@jocund Can you share the schema and/or typical records that can reproduce the issue?

jocund commented:
@akshayganeshen @fredwes Below is high level example table that is resulting in unknown type error. Would you know the cause? 

CREATE TABLE [db_dbo].[customer](
	[id] [uniqueidentifier] NOT NULL,
	[addresscity] [varchar](50) NULL,
	[addressline1] [nvarchar](50) COLLATE Latin1_General_BIN2 ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [<Your_Key>], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[addressstate] [varchar](2) NULL,
	[addresszip] [varchar](9) NULL,
	[First_name] [nvarchar](100) COLLATE Latin1_General_BIN2 ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [<Your_Key>], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[language] [varchar](2) NULL,
	[Last_name] [nvarchar](100) COLLATE Latin1_General_BIN2 ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [<Your_Key>], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[notes] [varchar](1000) NULL,
	[phonenumber] [nvarchar](20) COLLATE Latin1_General_BIN2 ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [<Your_Key>], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[companyname] [nvarchar](100) COLLATE Latin1_General_BIN2 ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [<Your_Key>], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[Submitted_by] [nvarchar](100) COLLATE Latin1_General_BIN2 ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [<Your_Key>], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[createdon] [uniqueidentifier] NOT NULL,
)

jocund commented:
@fredwes @akshayganeshen  Is this a defect? Please let me know

jonmifsud commented:
@fredwes this looks like some pretty solid work. 

We're helping out a client who has their app primarily in .net and we've done some work in Node. I'm trying to get this set up with Azure Key Vault and I think I'm missing some pieces.

I've managed to get the standard example encryption work just fine. However, I'm not sure what I'm missing to use the `ColumnEncryptionAzureKeyVaultProvider`, particularly when creating the table with fields. 

Would you still create the Master and Column encryption keys within the SQL query? I'm guessing at the very least the master key is obtained from the KeyVault however reading through the code I couldn't see which part/function I need to run to do this cleanly. 

Or is there a separate create table syntax when using Azure Key Vault

fredwes commented:
> @fredwes @akshayganeshen Is this a defect? Please let me know

@jocund I can't seem to reproduce the issue with the table schema you provided. Can you give me more information?

fredwes commented:
> @fredwes this looks like some pretty solid work.
> 
> We're helping out a client who has their app primarily in .net and we've done some work in Node. I'm trying to get this set up with Azure Key Vault and I think I'm missing some pieces.
> 
> I've managed to get the standard example encryption work just fine. However, I'm not sure what I'm missing to use the `ColumnEncryptionAzureKeyVaultProvider`, particularly when creating the table with fields.
> 
> Would you still create the Master and Column encryption keys within the SQL query? I'm guessing at the very least the master key is obtained from the KeyVault however reading through the code I couldn't see which part/function I need to run to do this cleanly.
> 
> Or is there a separate create table syntax when using Azure Key Vault

@jonmifsud The Master and Column encryption keys are created using separate SQL statements and then referenced when creating a table that uses them. You can see a basic example in my integration tests here: https://github.com/tediousjs/tedious/blob/66bd734c1dc14a1f9ec3d5d5f9308193a0ff2cf0/test/integration/always-encrypted-test.js

jonmifsud commented:
@fredwes thanks in the end we set up the Master and Column keys from within the management tool interface as it was easier... we are encrypting/decrypting data correctly so this works great.

v-tkhemasris commented:
When do you plan to merge this PR ?
